src/conversions/std/string.rs

```diff
@@ -233,7 +233,152 @@ impl FromPyObject<'_, '_> for char {
+    const OUTPUT_TYPE: &'static str = "str";
+
+    #[inline]
+    fn into_pyobject(self, py: Python<'py>) -> Result<Self::Output, Self::Error> {
+        // Convert cstr to &str, it's is safe because cstr is guaranteed to be valid UTF-8
+        Ok(PyString::new(py, self.to_str().unwrap()))
+    }
+
+    #[cfg(feature = "experimental-inspect")]
+    fn type_output() -> TypeInfo {
+        <String>::type_output()
+    }
+}
+
+impl<'py> IntoPyObject<'py> for std::ffi::CString {
+    type Target = PyString;
+    type Output = Bound<'py, Self::Target>;
+    type Error = Infallible;
+
+    #[cfg(feature = "experimental-inspect")]
+    const OUTPUT_TYPE: &'static str = "str";
+
+    #[inline]
+    fn into_pyobject(self, py: Python<'py>) -> Result<Self::Output, Self::Error> {
+        // Convert cstring to &str, it's safe because cstring is guaranteed to be valid UTF-8
+        Ok(PyString::new(py, self.to_str().unwrap()))
+    }
+
+    #[cfg(feature = "experimental-inspect")]
+    fn type_output() -> TypeInfo {
+        <String>::type_output()
+    }
+}
+
+impl<'py> IntoPyObject<'py> for &std::ffi::CString {
+    type Target = PyString;
+    type Output = Bound<'py, Self::Target>;
+    type Error = Infallible;
+
+    #[cfg(feature = "experimental-inspect")]
+    const OUTPUT_TYPE: &'static str = "str";
+
+    #[inline]
+    fn into_pyobject(self, py: Python<'py>) -> Result<Self::Output, Self::Error> {
+        Ok(PyString::new(py, self.to_str().unwrap()))
+    }
+
+    #[cfg(feature = "experimental-inspect")]
+    fn type_output() -> TypeInfo {
+        <String>::type_output()
+    }
+}
+
+/// Allows extracting cstr from Python objects.
+/// Accepts Python `str` objects and converts them to cstr.
+/// Fails if the string contains interior nul bytes.
+impl<'a> FromPyObject<'a, '_> for &'a std::ffi::CStr {
+    type Error = PyErr;
+
+    #[cfg(feature = "experimental-inspect")]
+    const INPUT_TYPE: &'static str = "str";
+
+    fn extract(obj: Borrowed<'a, '_, PyAny>) -> Result<Self, Self::Error> {
+        let py_string = obj.cast::<PyString>()?;
+
+        // Use PyUnicode_AsUTF8AndSize to get the raw bytes with guaranteed NUL termination
+        #[cfg(any(Py_3_10, not(Py_LIMITED_API)))]
+        {
+            let mut size: ffi::Py_ssize_t = 0;
+            let data: *const u8 =
+                unsafe { ffi::PyUnicode_AsUTF8AndSize(py_string.as_ptr(), &mut size).cast() };
+            if data.is_null() {
+                return Err(crate::PyErr::fetch(obj.py()));
+            }
+
+            // Create a slice that includes the NUL terminator
+            let bytes_with_nul = unsafe { std::slice::from_raw_parts(data, (size + 1) as usize) };
+
+            // Check for interior NUL bytes (excluding the final NUL)
+            for &byte in &bytes_with_nul[..size as usize] {
+                if byte == 0 {
+                    return Err(PyValueError::new_err("string contains interior NUL bytes"));
+                }
+            }
```

Comment:
No need to do this manually, we should make use of [CStr::from_bytes_with_nul](https://doc.rust-lang.org/stable/std/ffi/struct.CStr.html#method.from_bytes_with_nul) and just map the error.